lib/midcom/helper/imagepopup/handler/upload.php

```diff
@@ -0,0 +1,161 @@
+		reset ($_FILES);
+		$temp = current($_FILES);
+		
+		$accepted_origins = array("http://localhost", "http://openpsa/");
+		
+		if(is_uploaded_file($temp['tmp_name']))
+		{
+			if (isset($_SERVER['HTTP_ORIGIN']))
+			{
+				// same-origin requests won't set an origin. If the origin is set, it must be valid.
+				if (in_array($_SERVER['HTTP_ORIGIN'], $accepted_origins))
+				{
+					header('Access-Control-Allow-Origin: ' . $_SERVER['HTTP_ORIGIN']);
+				}
+				else
+				{
+					header("HTTP/1.0 403 Origin Denied");
+					return;
+				}
+			}
+		
+			if (preg_match("/([^\w\s\d\-_~,;:\[\]\(\).])|([\.]{2,})/", $temp['name']))
+			{
+				header("HTTP/1.0 500 Invalid file name.");
+				return;
+			}
+		
+			if (!in_array(strtolower(pathinfo($temp['name'], PATHINFO_EXTENSION)), array("gif", "jpg", "png")))
+			{
+				header("HTTP/1.0 500 Invalid extension.");
+				return;
+			}
+		}
+		
+		// Get the data
+		$temp_name = $temp['tmp_name'];
+		$mimetype = $temp['type'];
+		
+		switch($handler_id)
+		{
+			case '____ais-imagepopup-upload_image':
+				$parentguid = $args[0];
+				break;
+		}
+		
+		// Set modified filename 
+		$filename = $this->getModifyFilename($temp['name']);
+		
+		// Set the part of the location
+		$core = "/midcom-serveattachmentguid-";
+		
+		// Insert the image into database
+		$this->insertDatabase($filename, $mimetype, $parentguid);
+            
+        // Get guid of the image
+        $guid = $this->getDataFromDatabase($filename, "guid");
+        
+        // Get target to write the file
+        $target = $this->getDataFromDatabase($filename, "target");
+        
+        // Write the file
+        $this->writeTheFile($temp_name, $target);
+        
+        // Make a response for uploadImage() function
+        $location = $core. $guid . "/" . $filename;
+        
+        // Send a location with uploadImage() function
+		$this->_request_data['location'] =  $location;
+		
+		return true;
+	}
+
+	
+	public function _show_upload()
+	{
+		midcom_show_style('midcom_helper_imagepopup_upload');
+	}
+	
+	public function getModifyFilename($filename)
+	{
+		$pieces = explode('.', $filename);
+		$core = "modified";
+		$extension = array_pop($pieces);
+		$time = $this->getCurrentTime('d-m-Y-H-i-s');
+		$modifyFilename =  $core . $time . "." . $extension;
+	
+		return $modifyFilename;
+	}
+	
+	public function getCurrentTime($pattern)
+	{
+		return date($pattern, time());
+	}
+	
+	public function getDataFromDatabase($filename, $data)
+	{
+		$query = midcom_db_attachment::new_query_builder();
+		$query->add_constraint('name', '=', $filename);
+		$entry = $query->execute();
+		$guid = $entry[0]->guid;
+		
+		if($data == "guid") return $entry[0]->guid;
+		else if ($data == "target") return $entry[0];
+		
+		return false;
+	}
+	
+	public function writeTheFile($tmp, $target)
+	{
+		$source  = fopen($tmp, 'r');
+		if(!$source)
+		{
+			debug_add("Could not open file '{$this->_filename}' for reading", MIDCOM_LOG_ERROR);
+			return false;
```

Comment:
The return value from this function is never used, so if something goes wrong, the user doesn't get any information. Either throw an exception or check the return value and then show an error